server/container/util-java8/src/main/java/org/apache/james/util/CompletableFutureUtil.java

```diff
@@ -51,17 +58,29 @@
         return futurStream
             .thenCompose(stream ->
                 CompletableFutureUtil.allOf(
-                    stream.map(action::apply)));
+                    stream.map(action)));
     }
 
     public static <T, U> CompletableFuture<Stream<U>> map(CompletableFuture<Stream<T>> futurStream, Function<T, U> action) {
         return futurStream
             .thenApply(stream ->
-                stream.map(value ->
-                    action.apply(value)));
+                stream.map(action));
+    }
+
+    public static <T, U> CompletableFuture<Optional<T>> reduce(BinaryOperator<T> binaryOperator, CompletableFuture<Stream<T>> futureStream) {
+        return futureStream.thenApply(stream -> stream.reduce(binaryOperator));
     }
 
     public static <T> CompletableFuture<T> keepValue(Supplier<CompletableFuture<Void>> supplier, T value) {
         return supplier.get().thenApply(any -> value);
     }
+
+    public static <T> Function<Boolean, CompletableFuture<Boolean>> composeIfTrue(Supplier<CompletableFuture<T>> composeOperation) {
```

Comment:
I have personally no issue by this function. When you were speaking about monad helping with conditional @mbaechler what monad did you have in mind (Optional/Maybe) ?
But in the other hand maybe the place where you use this function a simple if or a ternary expression would be better. Because I have to look at the code of the function to be sure to understand what it was about, in the other end if I saw ` b ? something.thenApply(any -> b) : CompletableFuture.completedFuture(b)`.
I would immediately see what it was about.